File: ios/App/Shared/WidgetSnapshot.swift
import Foundation

/// What the Home Screen and Lock Screen widgets show, plus what Siri and
/// Spotlight know. The app writes it into the shared App Group whenever the
/// dashboard loads; the widget only reads.
struct WidgetSnapshot: Codable {
    /// The next programmed workout, e.g. "Upper A (Mon)". Nil without a program.
    var nextBout: String?
    /// Its template, for the widget's Start button.
    var nextTemplateId: String?
    var programName: String?
    var sessionsThisWeek: Int
    /// Workouts per week in the active program, when there is one.
    var sessionsPlanned: Int?
    var setsThisWeek: Int
    /// Monday of the week the numbers belong to (yyyy-MM-dd, the lifter's
    /// calendar). Past that week, the widget shows a fresh week instead.
    var weekStart: String
    /// Epoch milliseconds.
    var updatedAt: Double

    /// For Siri and Spotlight: the lifter's display unit ("kg" or "lb"),
    /// the days /log offers, and their bests (src/lib/actions/native.ts).
    var unit: String?
    var workouts: [Workout]?
    var lifts: [Lift]?
    /// This week's sets per muscle, in the dashboard chart's order.
    var muscles: [MuscleSets]?
    /// Estimated-max history per lift, most trained first.
    var trends: [Trend]?

    struct Workout: Codable {
        var templateId: String
        /// "Day 2: Upper".
        var label: String
    }

    struct Lift: Codable {
        /// The exercise id.
        var id: String
        var name: String
        /// The set behind the best estimated max, in the display unit.
        var bestWeight: Double
        var bestReps: Int
        var estimatedMax: Double
    }

    struct MuscleSets: Codable {
        var key: String
        var label: String
        var sets: Double
        /// One of the lifter's tracked weak points.
        var weak: Bool
    }

    struct Trend: Codable {
        /// The exercise id.
        var id: String
        var name: String
        var points: [Point]

        struct Point: Codable {
            /// yyyy-MM-dd.
            var date: String
            /// Best estimated 1RM that session, display unit.
            var e1rm: Double
        }
    }

    /// Starts the next programmed day (/log?start=…), for the widget's button.
    var startURL: URL? {
        nextTemplateId.flatMap { URL(string: "https://hellblazer.vercel.app/log?start=\($0)") }
    }

    static let appGroup = "group.com.kkrwhofrags.hellblazer"
    static let storageKey = "widget-snapshot"

    /// Stores a snapshot's JSON for the widgets (the app only).
    static func save(_ data: Data) {
        UserDefaults(suiteName: appGroup)?.set(data, forKey: storageKey)
    }

    static func load() -> WidgetSnapshot? {
        guard let data = UserDefaults(suiteName: appGroup)?.data(forKey: storageKey) else { return nil }
        return try? JSONDecoder().decode(WidgetSnapshot.self, from: data)
    }

    /// The first moment of the week after this snapshot's: when its numbers
    /// stop being "this week", and when the widget's timeline redraws.
    func weekEnd() -> Date? {
        let calendar = Calendar(identifier: .iso8601)
        let parser = DateFormatter()
        parser.calendar = calendar
        parser.locale = Locale(identifier: "en_US_POSIX")
        parser.dateFormat = "yyyy-MM-dd"
        guard let start = parser.date(from: weekStart) else { return nil }
        return calendar.date(byAdding: .day, value: 7, to: start)
    }

    func isFromPastWeek(now: Date = Date()) -> Bool {
        guard let end = weekEnd() else { return false }
        return now >= end
    }
}

File: src/lib/widget-snapshot.ts
import { format, parseISO, startOfISOWeek } from "date-fns";
import type { SupabaseClient } from "@supabase/supabase-js";
import type { Database } from "@/lib/database.types";
import { getProgramProgress, PROGRAM_SELECT, type ProgramWithDays } from "@/lib/data/programs";
import { dateInTimeZone } from "@/lib/local-date";
import { CHART_HIDDEN_MUSCLES, isWeakPoint, MUSCLE_CHART_ORDER, MUSCLE_LABEL, type Muscle } from "@/lib/muscles";
import type { WidgetSnapshot } from "@/lib/native-plugins";
import { toDisplayWeight, type Unit } from "@/lib/units";

/** The most-trained lifts Siri, Spotlight and the Lift Trend widget know about. */
const MAX_LIFTS = 40;
/** Estimated-max history for the Lift Trend widget: every lift it can pick, 16 sessions each. */
const TREND_LIFTS = MAX_LIFTS;
const TREND_POINTS = 16;

/**
 * Everything the iPhone app's widgets, Siri and Spotlight read, in one place
 * (ios/App/Shared/WidgetSnapshot.swift). Two callers:
 *
 * - the dashboard in the app (getNativeSnapshot), with the lifter's own client;
 * - /api/device/snapshot, which a silent push sends the phone to after a
 *   workout is finished anywhere, with the service role.
 *
 * So every query is scoped to `userId` explicitly, and the two functions it
 * calls take the lifter as an argument (honoured only without a signed-in user).
 * Weights are in the display unit; dates in the lifter's calendar.
 */
export async function buildWidgetSnapshot(
  db: SupabaseClient<Database>,
  { userId, unit, timeZone }: { userId: string; unit: Unit; timeZone: string },
): Promise<Omit<WidgetSnapshot, "updatedAt">> {
  const today = dateInTimeZone(new Date(), timeZone);
  const weekStart = format(startOfISOWeek(parseISO(today)), "yyyy-MM-dd");

  const [week, program, muscles, stats, trendRows] = await Promise.all([
    db
      .from("v_session_summary")
      .select("working_sets")
      .eq("user_id", userId)
      .gte("session_date", weekStart)
      .then(({ data, error }) => {
        if (error) throw error;
        return data ?? [];
      }),
    db
      .from("program")
      .select(PROGRAM_SELECT)
      .eq("user_id", userId)
      .eq("is_active", true)
      .order("position", { ascending: true, referencedTable: "program_day" })
      .maybeSingle()
      .then(({ data, error }) => {
        if (error) throw error;
        return (data as ProgramWithDays | null) ?? null;
      }),
    db
      .from("v_weekly_sets_per_muscle")
      .select("muscle, sets")
      .eq("user_id", userId)
      .eq("week", weekStart)
      .then(({ data, error }) => {
        if (error) throw error;
        return data ?? [];
      }),
    db.rpc("exercise_stats", { p_user: userId }).then(({ data, error }) => {
      if (error) throw error;
      return data ?? [];
    }),
    db
      .rpc("lift_trends", { p_user: userId, p_lifts: TREND_LIFTS, p_points: TREND_POINTS })
      .then(({ data, error }) => {
        if (error) throw error;
        return data ?? [];
      }),
  ]);

  const progress = program ? await getProgramProgress(program, { supabase: db, timeZone }) : null;
  const nextTemplate = progress && !progress.isCompleted ? progress.nextDay?.workout_template : null;

  // The days /log offers: the active program's, else every template.
  const workouts: NonNullable<WidgetSnapshot["workouts"]> = [];
  if (program) {
    const seen = new Set<string>();
    for (const day of program.program_day) {
      const t = day.workout_template;
      if (!t || seen.has(t.id)) continue;
      seen.add(t.id);
      workouts.push({ templateId: t.id, label: t.day_label || t.name });
    }
  } else {
    const { data, error } = await db
      .from("workout_template")
      .select("id, name, day_label")
      .eq("user_id", userId)
      .order("position", { ascending: true });
    if (error) throw error;
    for (const t of data ?? []) workouts.push({ templateId: t.id, label: t.day_label || t.name });
  }

  const setsByMuscle = new Map(muscles.map((m) => [m.muscle, Number(m.sets ?? 0)]));

  const trends = new Map<string, NonNullable<WidgetSnapshot["trends"]>[number]>();
  for (const row of trendRows) {
    const trend = trends.get(row.exercise_id) ?? { id: row.exercise_id, name: row.exercise_name, points: [] };
    trend.points.push({
      date: row.session_date,
      e1rm: Math.round(toDisplayWeight(Number(row.best_est_1rm), unit) * 10) / 10,
    });
    trends.set(row.exercise_id, trend);
  }

  const lifts = [...stats]
    .sort((a, b) => Number(b.sessions_logged) - Number(a.sessions_logged))
    .slice(0, MAX_LIFTS);
  const liftOrder = new Map(lifts.map((l, i) => [l.exercise_id, i]));

  return {
    nextBout: nextTemplate ? nextTemplate.day_label || nextTemplate.name : null,
    nextTemplateId: nextTemplate?.id ?? null,
    programName: program?.name ?? null,
    sessionsThisWeek: week.length,
    sessionsPlanned: progress?.daysPerWeek ?? null,
    setsThisWeek: week.reduce((n, s) => n + Number(s.working_sets ?? 0), 0),
    weekStart,
    unit,
    workouts,
    lifts: lifts.map((s) => ({
        id: s.exercise_id,
        name: s.exercise_name,
        bestWeight: toDisplayWeight(Number(s.best_weight_kg), unit),
        bestReps: Number(s.best_reps),
        estimatedMax: Math.round(toDisplayWeight(Number(s.best_est_1rm), unit)),
      })),
    muscles: MUSCLE_CHART_ORDER.filter((m) => !CHART_HIDDEN_MUSCLES.has(m)).map((m: Muscle) => ({
      key: m,
      label: MUSCLE_LABEL[m],
      sets: Math.round((setsByMuscle.get(m) ?? 0) * 10) / 10,
      weak: isWeakPoint(m),
    })),
    // Most-trained first: the Lift Trend widget shows the first until one's picked.
    trends: [...trends.values()].sort(
      (a, b) => (liftOrder.get(a.id) ?? Infinity) - (liftOrder.get(b.id) ?? Infinity),
    ),
  };
}
